apps/api-extractor/src/api/ExtractorConfig.ts

```diff
@@ -856,42 +888,81 @@ export class ExtractorConfig {
         }
       }
 
-      let apiReportEnabled: boolean = false;
-      let reportFilePath: string = '';
-      let reportTempFilePath: string = '';
-      let apiReportIncludeForgottenExports: boolean = false;
-      if (configObject.apiReport) {
-        apiReportEnabled = !!configObject.apiReport.enabled;
-
-        const reportFilename: string = ExtractorConfig._expandStringWithTokens(
-          'reportFileName',
-          configObject.apiReport.reportFileName || '',
-          tokenContext
-        );
+      const apiReportEnabled: boolean = configObject.apiReport?.enabled ?? false;
+      const apiReportIncludeForgottenExports: boolean =
+        configObject.apiReport?.includeForgottenExports ?? false;
+      let reportDirectoryPath: string = tokenContext.projectFolder;
+      let reportTempDirectoryPath: string = tokenContext.projectFolder;
+      const reportConfigs: IApiReportConfig[] = [];
+      if (apiReportEnabled) {
+        // Undefined case checked above where we assign `apiReportEnabled`
+        const apiReportConfig: IConfigApiReport = configObject.apiReport!;
+
+        let reportFileNameBase: string;
+
+        // The config spec allows the user to optionally include an extension postfix in the report file name.
```

Comment:
It was somewhat ambiguous before. It was documented as being required, but the code didn't validate this. I believe it was happy to let you use any extension or no extension.

I figured for strict backwards compat, maintaining the flexibility to some degree made sense, and I updated the docs to reflect the implemented contract.

If we would prefer a stricter contract (with enforcement), I can change things here. That said, it seems odd to me to require someone specify a specific file extension. If we only support one kind of file output, we can generate that for them. 🤷‍♂️

Reply:
> If we would prefer a stricter contract (with enforcement), I can change things here. 

I'm kind of in favor of that, for two reasons:

1. The configuration is already a bit confusing:  This setting originally "required" the file extension because `"reportFileName": "<unscopedPackageName>.api.md"` seemed more intuitive than `"reportMeaninglessPathFragment": "<unscopedPackageName>"` for a person who is scanning the config file trying to determine how `my-library.api.md` got named.  But now we are introducing a transform that converts that string to `my-library.public.api.md`, and in fact the original name `my-library.api.md` may not get written to disk at all (if its variant is unused).  So the situation is already not super intuitive. (It really should be something like `"<unscopedPackageName>.<variant>.api.md"` but that has some other tradeoffs, e.g. do we get `..` with the default variant, etc.)

   Therefore it seems the system will be more learnable and less error-prone if we are strict about the input format.

2. `.api.md` is our standard file extension. Is there a compelling reason why someone needs to create PDF files with `.fdp` file extension or PNG files without the file extension? It seems like a somewhat unlikely requirement.

Reply:
> 1. The configuration is already a bit confusing:  This setting originally "required" the file extension because `"reportFileName": "<unscopedPackageName>.api.md"` seemed more intuitive than `"reportMeaninglessPathFragment": "<unscopedPackageName>"` for a person who is scanning the config file trying to determine how `my-library.api.md` got named.  But now we are introducing a transform that converts that string to `my-library.public.api.md`, and in fact the original name `my-library.api.md` may not get written to disk at all (if its variant is unused).  So the situation is already not super intuitive. (It really should be something like `"<unscopedPackageName>.<variant>.api.md"` but that has some other tradeoffs, e.g. do we get `..` with the default variant, etc.)
>    Therefore it seems the system will be more learnable and less error-prone if we are strict about the input format.

I guess I personally don't find the idea of "reportFileName" being consumed as an extension-less "fragment" to be all too confusing, especially if we discourage or disallow specifying the extension pattern (though this would be a breaking change).

I'm not opposed to re-adding the `.api.md` requirement, but if we add enforcement for it we would be introducing a functional breaking change, despite the existing documented contract. But as you mentioned, this almost seems more confusing given that we would be splicing in the `<variant>` string. The `<unscopedPackageName>.<variant>.api.md` idea is interesting, but I agree that there is some ambiguity there.

All of this leads me to think that the ideal would be to _not_ recommend providing the `.api.md` postfix, and document how the extensionless name is consumed. Then we will _always_ output files ending in `.api.md`, which I would argue is less confusing than asking for a specific extension that we would always want to output anyways. For_ backwards compatibility, we can continue to support the `.api.md` postfix when we see it, but probably leave that support undocumented to encourage users to just specify the extension-less name component.

Curious if you feel strongly otherwise here.

> 2. `.api.md` is our standard file extension. Is there a compelling reason why someone needs to create PDF files with `.fdp` file extension or PNG files without the file extension? It seems like a somewhat unlikely requirement.

No, my point above was just that the code didn't strictly enforce that the extension was `.api.md`. The docs asked for it, but the code would happily use whatever the user specified. So if they did specify `foo.pdf`, we would generate that file with markdown formatted contents.

I agree that we only output markdown, so we should always output `.md` files. This is why I find it odd that ask the user to specify the extension at all. I get your point about `reportMeaninglessPathFragment`, but I think calling it `reportFileName` and documenting what that means and how it gets consumed is a pretty reasonable compromise that users will be able to understand pretty easily. If there is any ambiguity (or if they don't read the docs), that will be clarified after they run the tool for the first time and see the output.

Reply:
Just to clarify, your proposal is:

1. Update the docs everywhere to stipulate that `"reportFileName"` should _never_ include a file extension.  It will be automatically appended, as well as a suffix like `.public.api.md` for variants.
2. However, if the filename happens have the specific extension `.api.md` then it will be silently removed (and re-added) to avoid breaking backwards compatibility with existing config files.

This can work, although we will forevermore have config files everywhere with `"reportFileName": "thing.api.md"` which might undermine that definition.

(Another idea would be to introduce a new setting like `"reportName": "<unscopedPackageName>"`, deprecating the old setting and planning to remove it in the next major release, and requiring the new setting if the variants feature is used.  However, this idea gets messy when we consider `"$extends"` and handling of default values.  So I think your design is maybe superior.)

And what if someone really really needs a different file extension? Then we could always introduce a `reportFileExtension` setting.

One final consideration is why is this designed differently from the quite straightforward `dtsRollup.untrimmedFilePath`, `dtsRollup.betaTrimmedFilePath`, `dtsRollup.publicTrimmedFilePath` settings?  I guess the answer is that we expect there to be a small number of trimming possibilities compared to variant profiles?

Reply:
> One final consideration is why is this designed differently from the quite straightforward `dtsRollup.untrimmedFilePath`, `dtsRollup.betaTrimmedFilePath`, `dtsRollup.publicTrimmedFilePath` settings? I guess the answer is that we expect there to be a small number of trimming possibilities compared to variant profiles?

Actually I remember the answer to this question now:

- For .d.ts rollups, the user often needs fairly precise control over where each file is written and what is its name, because they can be import paths that are part of a package's API contract.
- Whereas for .api.md reports, I had been thinking about a potential mistake where e.g. the `alpha` variant was being generated but the latest configuration removed that, and so `my-library.alpha.api.md` accidentally is being tracked by Git but no longer getting generated.  To prevent such confusion, we could introduce a feature something like `disallowUnusedVariantFiles` where API Extractor could scan the target folder for such files, which requires more strict control about how they are named.

With these thoughts, I'm 👍 👍 for your proposal.

Reply:
> This can work, although we will forevermore have config files everywhere with "reportFileName": "thing.api.md" which might undermine that definition.

At least until the next major version rev, yes. We could basically note that providing a file extension is deprecated.

Do the maintainers of this repo have a place where they track such follow-up work? I'm happy to add something there as needed.

Reply:
Updated the code and docs to match the following contract:
* Don't specify extension
* `reportFileName` is enterpreted as the body component of `<body>.<variant>.api.md` _except_ we will (as a special case) search for and strip out `.api.md` if we see it. This special-casing should be removed in version 8.